add a card footer in gallery containing meta data and link to single post view

Fensterbank opened the pull request:
I moved and renamed the class hovercard_footer to the base.scss as 'card_footer' to make it more generic for different use cases.
This pull request closes diaspora/diaspora#6610 and closes #6236

![diaspora-gallery](https://cloud.githubusercontent.com/assets/1684826/12143057/65428ae6-b47d-11e5-93fc-bbb36a5010a1.jpg)

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -23,6 +23,13 @@ class Photo < ActiveRecord::Base
         :width => photo.width }
     }, :as => :dimensions
+    t.add lambda { |photo|
+      { :id => photo.status_message.id,
```
Space inside { detected.

diaspora-code-review replied:
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -23,6 +23,13 @@ class Photo < ActiveRecord::Base
     }, :as => :dimensions
+    t.add lambda { |photo|
+      { :id => photo.status_message.id,
+        :likes_count => photo.status_message.likes_count,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review replied:
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -23,6 +23,13 @@ class Photo < ActiveRecord::Base
+    t.add lambda { |photo|
+      { :id => photo.status_message.id,
+        :likes_count => photo.status_message.likes_count,
+        :comments_count => photo.status_message.comments_count,
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -23,6 +23,13 @@ class Photo < ActiveRecord::Base
+      { :id => photo.status_message.id,
+        :likes_count => photo.status_message.likes_count,
+        :comments_count => photo.status_message.comments_count,
+        :reshares_count => photo.status_message.reshares_count
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/models/photo.rb:
```diff
@@ -23,6 +23,13 @@ class Photo < ActiveRecord::Base
+        :comments_count => photo.status_message.comments_count,
+        :reshares_count => photo.status_message.reshares_count
+      }
+    }, :as => :status_message
```
Use the new Ruby 1.9 hash syntax.

diaspora-code-review commented on app/presenters/photo_presenter.rb:
```diff
@@ -10,6 +10,12 @@ def base_hash
         large: url(:scaled_full)
+      },
+      status_message: {
+        id: status_message.id,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review commented on app/presenters/photo_presenter.rb:
```diff
@@ -10,6 +10,12 @@ def base_hash
+      },
+      status_message: {
+        id: status_message.id,
+        likes_count: status_message.likes_count,
```
Align the elements of a hash literal if they span more than one line.

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
+  font-size: 10px;
+  line-height: 18px;
+
+  border-top: 1px solid #cccccc;
```
Color literals like `#cccccc` should only be used in variable declarations; they should be referred to via variable everywhere else.

diaspora-code-review replied:
Color `#cccccc` should be written as `#ccc`

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
+  bottom: 0;
+  left: 0;
+  background-color: $background-grey;
+  margin-left: 0 !important;
```
!important should not be used

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
+
+// grey card footer
+.card_footer {
+  position: absolute;
```
Properties should be ordered background-color, border-top, bottom, font-size, left, line-height, margin-left, min-height, position, width

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
+
+    a {
+      color: $text-grey;
+      margin-right: 4px;
```
Properties should be ordered color, font-weight, margin-right

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
 }
+
+// grey card footer
+.card_footer {
```
Selector `card_footer` should be written in lowercase with hyphens

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,30 @@ pre { word-wrap: break-word; }
+
+  border-top: 1px solid #cccccc;
+
+  .footer_container {
```
Selector `footer_container` should be written in lowercase with hyphens

diaspora-code-review commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -15,3 +18,8 @@
 .media .img img { display: block }
 .media .imgEt { float: right; margin-left: 10px; }
 
+.media .gallery-picture img { margin-bottom: 20px; }
```
Merge rule `.media .gallery-picture img` with rule on line 1

diaspora-code-review commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -15,3 +18,8 @@
 
+.media .gallery-picture img { margin-bottom: 20px; }
+
+.media .card_footer .footer_container {  font-size: 13px; }
```
Merge rule `.media .card_footer .footer_container` with rule on line 1

diaspora-code-review replied:
Selector `card_footer` should be written in lowercase with hyphens

diaspora-code-review replied:
Selector `footer_container` should be written in lowercase with hyphens

diaspora-code-review commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -15,3 +18,8 @@
+
+.media .card_footer .footer_container {  font-size: 13px; }
+
+.media .card_footer span { float: right; }
```
Merge rule `.media .card_footer span` with rule on line 1

diaspora-code-review replied:
Selector `card_footer` should be written in lowercase with hyphens

Flaburgan commented:
Thanks for this work! Can you please add a screenshot of how it looks?
Btw you can check pronto errors locally by running `bin/pronto run -c upstream/develop`

Fensterbank commented:
I'm fixing most of the pronto errors atm and will add a screenshot soon.

diaspora-code-review commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,27 @@ pre { word-wrap: break-word; }
+  font-size: 10px;
+  left: 0;
+  line-height: 18px;
+  margin-left: 0 !important;
```
!important should not be used

diaspora-code-review commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -1,4 +1,20 @@
+    margin-right: 10px;
+  }
+
+  .img img { display: block }
```
Merge rule `.img img` with rule on line 11

diaspora-code-review replied:
Declaration should be terminated by a semicolon

Fensterbank commented:
I added a screenshot to the PR description above and fixed the pronto errors.
Sorry, I needed a few commits because each commit resulted in a few new errors.

Fensterbank committed:
add a card footer in gallery containing meta data and link to single post view

jhass commented:
See https://wiki.diasporafoundation.org/Styleguide#Automatic_local_review

Please squash the commits together after review is done.

Flaburgan commented:
I'm not too sure about the like / comment / reshare counts actually: in your design, we think that it refers to the picture. But in fact, they are referring to the post. A post often contains several images, that means all the image will have the same footer. I don't know but there is something confusing there.

Fensterbank commented:
@jhass: I squashed the commits.
@Flaburgan: Yes, that's right. Cause picture specific likes/comments does not exist in Diaspora, the counts are referring to the post and multiple images would have the same footer referring to the same single post view.

It could be confusing, but it's better without having any reference to the post like it was before.
Changing this logic, e.g. by summarize multiple pictures from a single post into a single thumbnail card with one footer, would be a possible way, but needs many code and layout changes in the gallery page.
It's too much for me at the moment, but maybe something we can think about in the future?

Flaburgan commented:
@Fensterbank Maybe a mention "This post received ..." or something like that would be enough to disambiguate

Fensterbank commented:
@Flaburgan I'm afraid we don't have the place for this on lower resoultions. :-1: 
![bildschirmfoto am 2016-01-06 um 15 02 30](https://cloud.githubusercontent.com/assets/1684826/12144263/c6a131f8-b486-11e5-9162-8a2fbbb3988b.png)

The only way I know at the moment to disambiguate it without much work is to not show the counts, but then we lost information we could show there.

jhass commented:
I think I don't think it's useful information there to be honest, it's more confusing than helping anything. If anything, link the post or group the pictures by post and show the information for the group.

Fensterbank commented:
Okay, I just removed the count values of likes, comments and reshares.
The footer is displayed with the default created_at link (which is already known by the default stream view) which redirects to the single post view.
It's an improvement and fixes #6610 by giving a link with a well-known behaviour without adding unnecessary or confusing information.
This should be good to merge the pull request.

![diaspora-gallery1](https://cloud.githubusercontent.com/assets/1684826/12145537/00eb12fe-b48f-11e5-9ba1-ea47b41fb33c.jpg)

The count values can be added again some day in a new PR, when I figured out a good way to group the pictures of a post in the gallery view.

jhass commented on app/assets/templates/photo_tpl.jst.hbs:
```diff
@@ -26,4 +26,11 @@
   </div>
+  <div class="card-footer">
+    <div class="footer-container">
+      <a href="/posts/{{status_message.id}}">
```
`{{urlTo "posts" status_message.id}}`

Flaburgan commented on app/presenters/photo_presenter.rb:
```diff
@@ -4,12 +4,18 @@ def base_hash
+      },
+      status_message: {
+        id:             status_message.id,
+        likes_count:    status_message.likes_count,
```
You removed this information, you probably don't need it anymore.

Fensterbank replied:
I just removed the unused properties in photo.rb. I forgot it in photo_presenter.rb, but the photo_presenter.rb isn't used anywhere in the code. ;)

jhass replied:
Which is only one more reason to not touch it (and maybe do a *separate* PR for putting it to use or dropping it entirely).

Flaburgan commented:
OK I tested this locally and it looks good to me :+1:

svbergerem commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,27 @@ pre { word-wrap: break-word; }
+  margin-left: 0;
+  min-height: 19px;
+  position: absolute;
+  width: 100%;
```
For the sake of unification I think I'd prefer
```diff
diff --git a/app/assets/stylesheets/base.scss b/app/assets/stylesheets/base.scss
index d5b4f3b..59a40cf 100644
--- a/app/assets/stylesheets/base.scss
+++ b/app/assets/stylesheets/base.scss
@@ -80,11 +80,11 @@ pre { word-wrap: break-word; }
   background-color: $background-grey;
   border-top: 1px solid $border-medium-grey;
   bottom: 0;
-  font-size: 10px;
+  font-size: $font-size-small;
   left: 0;
-  line-height: 18px;
+  line-height: $line-height-computed;
   margin-left: 0;
-  min-height: 19px;
+  min-height: $line-height-computed + 1;
   position: absolute;
   width: 100%;
```

svbergerem commented on app/assets/stylesheets/base.scss:
```diff
@@ -74,3 +74,27 @@ pre { word-wrap: break-word; }
+      margin-right: 4px;
+    }
+  }
+}
```
Maybe move the card-footer stuff to a new file?

svbergerem commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -1,4 +1,20 @@
+
+  .card-footer .footer-container {  font-size: 13px; }
+
+  .card-footer span { float: right; }
```
Again, you could move these card-footer / gallery-picture specific lines to a dedicated card-footer file.

Fensterbank replied:
the gallery-picture specific lines belongs to this file I think, because all I know media/media-box is only used for the gallery.

svbergerem replied:
We use those Bootstrap media objects for different kinds of elements:
```
$ grep -Rn "\"media" app/assets/templates
app/assets/templates/comment-stream_tpl.jst.hbs:3:    <div class="media">
app/assets/templates/contact_tpl.jst.hbs:6:  <div class="media-object pull-left">
app/assets/templates/contact_tpl.jst.hbs:10:  <div class="media-body">
app/assets/templates/likes-info_tpl.jst.hbs:3:    <div class="media">
app/assets/templates/reshares-info_tpl.jst.hbs:3:    <div class="media">
app/assets/templates/stream-element_tpl.jst.hbs:1:<div class="media {{#if showPost}} {{#if nsfw}} shield-off {{/if}} {{else}} shield-active {{/if}}">
app/assets/templates/reshare_tpl.jst.hbs:2:  <div class="media">
app/assets/templates/photo_tpl.jst.hbs:1:<div class="media">
```
(the output of `grep -Rn "\.media" app/views` is even longer)

and as far as I can tell you have to add the margin to have enough space for the card-footer.

svbergerem commented:
I added some remarks. Apart from those this is looking good. :)

svbergerem commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -1,4 +1,20 @@
+
+  .gallery-picture img { margin-bottom: 20px; }
+
+  .card-footer .footer-container {  font-size: 13px; }
```
Please use `$font-size-base` instead.

svbergerem commented on app/assets/stylesheets/media-box.scss:
```diff
@@ -1,4 +1,20 @@
+  margin: 10px;
+  position: relative;
+
+  .gallery-picture img { margin-bottom: 20px; }
```
I guess this would be `$line-height-computed + 2;` but I didn't test that.

Fensterbank commented:
Thanks for the review!
You're right, I should replace the absolute values with existing variables.

Also I will make a new card-footer.scss containing all card-footer specific styles.

denschub commented:
Ping!

diaspora-code-review commented on app/assets/stylesheets/card-footer.scss:
```diff
@@ -0,0 +1,30 @@
+  width: 100%;
+
+  .footer-container {
+    padding: 1px 5px;
```
Properties should be ordered font-size, padding

Fensterbank replied:
this seems to be from an outdated diff. I removed the wrong font-size line here a few minutes ago. Pronto doesn't say anything locally.

Fensterbank commented:
Sorry, I was a bit short in time. I implemented the changes @svbergerem suggested and hope this PR can now be merged without further changes.

svbergerem commented:
Thank you!
